In [1]:
import pandas as pd
import re

In [2]:
st2017 = pd.read_csv('../data/OEWS/2017state_M_dl.csv')
st2017.head(3)

AREA  ST    STATE OCC_CODE               OCC_TITLE OCC_GROUP    TOT_EMP  \
0     1  AL  Alabama  00-0000         All Occupations     total  1,922,570   
1     1  AL  Alabama  11-0000  Management Occupations     major     69,950   
2     1  AL  Alabama  11-1011        Chief Executives  detailed      1,120   

  EMP_PRSE JOBS_1000 LOC_Q  ... H_MEDIAN H_PCT75 H_PCT90 A_PCT10  A_PCT25  \
0      0.4  1000.000  1.00  ...    15.77   25.01   37.83  17,770   21,740   
1      1.0    36.385  0.71  ...    46.63   64.12   90.80  52,130   70,130   
2      6.0     0.585  0.40  ...    94.45       #       #  84,520  125,290   

  A_MEDIAN  A_PCT75  A_PCT90 ANNUAL HOURLY  
0   32,800   52,020   78,690    NaN    NaN  
1   96,980  133,360  188,860    NaN    NaN  
2  196,460        #        #    NaN    NaN  

[3 rows x 25 columns]

In [3]:
st2017 = st2017[['AREA', 'ST', 'STATE','OCC_CODE', 'OCC_TITLE', 'TOT_EMP', 'JOBS_1000', 'LOC_Q', 'A_MEAN', 
                         'A_PCT10', 'A_PCT25', 'A_MEDIAN', 'A_PCT75', 'A_PCT90']]
st2017.loc[:, 'year'] = 2017
st2017.head(2)

AREA  ST    STATE OCC_CODE               OCC_TITLE    TOT_EMP JOBS_1000  \
0     1  AL  Alabama  00-0000         All Occupations  1,922,570  1000.000   
1     1  AL  Alabama  11-0000  Management Occupations     69,950    36.385   

  LOC_Q   A_MEAN A_PCT10 A_PCT25 A_MEDIAN  A_PCT75  A_PCT90  year  
0  1.00   43,170  17,770  21,740   32,800   52,020   78,690  2017  
1  0.71  111,150  52,130  70,130   96,980  133,360  188,860  2017

In [4]:
col2017 = pd.Series(['AREA', 'ST', 'STATE','OCC_CODE', 'OCC_TITLE', 'TOT_EMP', 'JOBS_1000', 'LOC_Q', 'A_MEAN', 
                         'A_PCT10', 'A_PCT25', 'A_MEDIAN', 'A_PCT75', 'A_PCT90', 'year'])
lowcol2017 = col2017.str.lower()
st2017.columns = lowcol2017
st2017.head(3)

area  st    state occ_code               occ_title    tot_emp jobs_1000  \
0     1  AL  Alabama  00-0000         All Occupations  1,922,570  1000.000   
1     1  AL  Alabama  11-0000  Management Occupations     69,950    36.385   
2     1  AL  Alabama  11-1011        Chief Executives      1,120     0.585   

  loc_q   a_mean a_pct10  a_pct25 a_median  a_pct75  a_pct90  year  
0  1.00   43,170  17,770   21,740   32,800   52,020   78,690  2017  
1  0.71  111,150  52,130   70,130   96,980  133,360  188,860  2017  
2  0.40  207,090  84,520  125,290  196,460        #        #  2017

In [5]:
bad1 = ['*', '**', '#']
good1 = [0, 0, 145601]
st2017 = st2017.replace(to_replace=bad1, value=good1)
st2017.head(3)

area  st    state occ_code               occ_title    tot_emp jobs_1000  \
0     1  AL  Alabama  00-0000         All Occupations  1,922,570  1000.000   
1     1  AL  Alabama  11-0000  Management Occupations     69,950    36.385   
2     1  AL  Alabama  11-1011        Chief Executives      1,120     0.585   

  loc_q   a_mean a_pct10  a_pct25 a_median  a_pct75  a_pct90  year  
0  1.00   43,170  17,770   21,740   32,800   52,020   78,690  2017  
1  0.71  111,150  52,130   70,130   96,980  133,360  188,860  2017  
2  0.40  207,090  84,520  125,290  196,460   145601   145601  2017

In [6]:
occ_codes_2017 = ['11-3021', '13-1081', '13-1111', '13-1161', '15-1111', '15-1121', '15-1122', '15-1131', '15-1132', '15-1133',
                '15-1134', '15-1141', '15-1142', '15-1143', '15-1151', '15-1152', '15-1199', '15-2031', '15-2041', '17-2061',
                '43-9011', '43-9021', '43-9111', '49-2011']
st2017 = st2017.loc[st2017.occ_code.isin(occ_codes_2017)].reset_index(drop=True)

print(st2017.shape)
st2017.tail()

(1224, 15)


area  st           state occ_code  \
1219    78  VI  Virgin Islands  15-1121   
1220    78  VI  Virgin Islands  15-1142   
1221    78  VI  Virgin Islands  15-1151   
1222    78  VI  Virgin Islands  15-1152   
1223    78  VI  Virgin Islands  43-9021   

                                        occ_title tot_emp jobs_1000 loc_q  \
1219                    Computer Systems Analysts      50     1.344  0.33   
1220  Network and Computer Systems Administrators      40     1.128  0.43   
1221            Computer User Support Specialists     100     2.670  0.62   
1222         Computer Network Support Specialists      50     1.203  0.92   
1223                            Data Entry Keyers     100     2.793  2.21   

      a_mean a_pct10 a_pct25 a_median a_pct75  a_pct90  year  
1219  51,710  34,610  40,670   48,510  62,030   75,870  2017  
1220  75,750  46,860  55,560   65,320  84,630  142,960  2017  
1221  43,340  31,140  34,500   39,990  49,650   60,760  2017  
1222  43,070  20,690  33,720   44,360  52,250   60,870  2017  
1223  31,760  19,730  24,180   31,000  38,780   46,000  2017

In [7]:
st2017 = st2017.loc[~st2017.state.isin(['Guam', 'Virgin Islands'])].reset_index(drop=True)
st2017


area  st        state occ_code  \
0        1  AL      Alabama  11-3021   
1        1  AL      Alabama  13-1081   
2        1  AL      Alabama  13-1111   
3        1  AL      Alabama  13-1161   
4        1  AL      Alabama  15-1121   
...    ...  ..          ...      ...   
1200    72  PR  Puerto Rico  17-2061   
1201    72  PR  Puerto Rico  43-9011   
1202    72  PR  Puerto Rico  43-9021   
1203    72  PR  Puerto Rico  43-9111   
1204    72  PR  Puerto Rico  49-2011   

                                              occ_title tot_emp jobs_1000  \
0             Computer and Information Systems Managers   3,120     1.624   
1                                          Logisticians   3,590     1.868   
2                                   Management Analysts   4,250     2.209   
3     Market Research Analysts and Marketing Special...   3,370     1.754   
4                             Computer Systems Analysts   4,010     2.084   
...                                                 ...     ...       ...   
1200                        Computer Hardware Engineers     250     0.284   
1201                                 Computer Operators     830     0.945   
1202                                  Data Entry Keyers   2,260     2.577   
1203                             Statistical Assistants     290     0.329   
1204  Computer, Automated Teller, and Office Machine...     540     0.612   

     loc_q   a_mean a_pct10 a_pct25 a_median  a_pct75  a_pct90  year  
0     0.63  123,790  66,710  90,290  116,990  149,100  184,950  2017  
1     1.67   88,450  53,510  69,880   88,680  105,450  121,150  2017  
2     0.48   93,470  43,590  59,540   81,290  106,170  152,360  2017  
3     0.42   59,210  29,260  39,940   53,250   73,590   98,220  2017  
4     0.51   81,960  53,510  64,190   78,800   97,710  117,970  2017  
...    ...      ...     ...     ...      ...      ...      ...   ...  
1200  0.61   66,600  40,220  49,990   62,720   83,410   99,740  2017  
1201  3.33   22,140  16,320  17,360   19,090   24,200   32,100  2017  
1202  2.04   21,090  16,210  17,200   18,840   23,150   30,540  2017  
1203  4.64   26,240  19,550  22,280   25,890   29,090   31,060  2017  
1204  0.87   29,230  19,140  22,210   28,440   36,030   40,490  2017  

[1205 rows x 15 columns]

### 2017 State data should have 1205 rows, 50 states plus District of Columbia and Puerto Rico.

In [8]:
st2010 = pd.read_csv('../data/OEWS/2010state_M_dl.csv')
st2010.head(3)

AREA  ST    STATE OCC_CODE               OCC_TITLE  GROUP    TOT_EMP  \
0     1  AL  Alabama  00-0000         All Occupations  total  1,807,480   
1     1  AL  Alabama  11-0000  Management Occupations  major     69,180   
2     1  AL  Alabama  11-1011        Chief Executives    NaN      1,750   

  EMP_PRSE JOBS_1000 LOC QUOTIENT  ... H_MEDIAN H_PCT75 H_PCT90 A_PCT10  \
0      0.4  1000.000        1.000  ...    14.21   22.57   33.50  16,470   
1      1.0    38.274        0.808  ...    40.04   56.44       #  44,140   
2      4.8     0.971        0.451  ...        #       #       #  85,270   

   A_PCT25 A_MEDIAN  A_PCT75 A_PCT90 ANNUAL HOURLY  
0   19,630   29,570   46,940  69,670    NaN    NaN  
1   60,540   83,290  117,390       #    NaN    NaN  
2  120,370        #        #       #    NaN    NaN  

[3 rows x 25 columns]

In [9]:
st2010 = st2010[['AREA', 'ST', 'STATE','OCC_CODE', 'OCC_TITLE', 'TOT_EMP', 'JOBS_1000', 'LOC QUOTIENT', 'A_MEAN', 
                         'A_PCT10', 'A_PCT25', 'A_MEDIAN', 'A_PCT75', 'A_PCT90']]
st2010.loc[:, 'year'] = 2010
st2010.head(2)

AREA  ST    STATE OCC_CODE               OCC_TITLE    TOT_EMP JOBS_1000  \
0     1  AL  Alabama  00-0000         All Occupations  1,807,480  1000.000   
1     1  AL  Alabama  11-0000  Management Occupations     69,180    38.274   

  LOC QUOTIENT  A_MEAN A_PCT10 A_PCT25 A_MEDIAN  A_PCT75 A_PCT90  year  
0        1.000  38,590  16,470  19,630   29,570   46,940  69,670  2010  
1        0.808  96,480  44,140  60,540   83,290  117,390       #  2010

In [10]:
bad1 = ['*', '**', '#']
good1 = [0, 0, 145601]
st2010 = st2010.replace(to_replace=bad1, value=good1)
st2010.head(2)

AREA  ST    STATE OCC_CODE               OCC_TITLE    TOT_EMP JOBS_1000  \
0     1  AL  Alabama  00-0000         All Occupations  1,807,480  1000.000   
1     1  AL  Alabama  11-0000  Management Occupations     69,180    38.274   

  LOC QUOTIENT  A_MEAN A_PCT10 A_PCT25 A_MEDIAN  A_PCT75 A_PCT90  year  
0        1.000  38,590  16,470  19,630   29,570   46,940  69,670  2010  
1        0.808  96,480  44,140  60,540   83,290  117,390  145601  2010

In [11]:
occ_codes_2010 = ['11-3021', '13-1081', '13-1111', '13-1161', '15-1111', '15-1121', '15-1122', '15-1131', '15-1132', '15-1133',
                '15-1134', '15-1141', '15-1142', '15-1143', '15-1151', '15-1152', '15-1199', '15-2031', '15-2041', '17-2061',
                '43-9011', '43-9021', '43-9111', '49-2011']
st2010 = st2010.loc[st2010.OCC_CODE.isin(occ_codes_2010)].reset_index(drop=True)
st2010 = st2010.loc[~st2010.STATE.isin(['Guam', 'Virgin Islands'])].reset_index(drop=True)
print(st2010.shape)
st2010.tail()

(902, 15)


AREA  ST        STATE OCC_CODE  \
897    72  PR  Puerto Rico  17-2061   
898    72  PR  Puerto Rico  43-9011   
899    72  PR  Puerto Rico  43-9021   
900    72  PR  Puerto Rico  43-9111   
901    72  PR  Puerto Rico  49-2011   

                                             OCC_TITLE TOT_EMP JOBS_1000  \
897                        Computer Hardware Engineers     240     0.251   
898                                 Computer Operators     860     0.907   
899                                  Data Entry Keyers   2,720     2.857   
900                             Statistical Assistants     750     0.784   
901  Computer, Automated Teller, and Office Machine...     910     0.954   

    LOC QUOTIENT  A_MEAN A_PCT10 A_PCT25 A_MEDIAN A_PCT75 A_PCT90  year  
897        0.476  59,870  35,640  44,170   57,230  73,090  92,090  2010  
898        1.399  22,100  15,800  17,010   19,480  25,570  31,200  2010  
899        1.654  19,400  15,550  16,110   17,820  20,560  26,950  2010  
900        6.433  23,200  15,830  17,510   22,890  27,550  30,160  2010  
901        1.099  25,470  16,360  19,010   23,330  29,670  36,400  2010

In [12]:
col2010 = pd.Series(['AREA', 'ST', 'STATE','OCC_CODE', 'OCC_TITLE', 'TOT_EMP', 'JOBS_1000', 'LOC QUOTIENT', 'A_MEAN', 
                         'A_PCT10', 'A_PCT25', 'A_MEDIAN', 'A_PCT75', 'A_PCT90', 'year'])
lowcol2010 = col2010.str.lower()
st2010.columns = lowcol2010
st2010 = st2010.rename(columns= {'loc quotient': 'loc_q'})
st2010.head(3)

area  st    state occ_code                                  occ_title  \
0     1  AL  Alabama  11-3021  Computer and Information Systems Managers   
1     1  AL  Alabama  13-1081                               Logisticians   
2     1  AL  Alabama  13-1111                        Management Analysts   

  tot_emp jobs_1000  loc_q   a_mean a_pct10 a_pct25 a_median  a_pct75  \
0   2,390     1.321  0.582  107,250  68,350  81,540  103,280  128,600   
1   3,350     1.855  2.250   80,020  48,230  62,220   80,280   95,150   
2   4,400     2.436  0.577   83,690  43,780  57,180   74,590   98,950   

   a_pct90  year  
0  149,220  2010  
1  108,120  2010  
2  129,770  2010

In [13]:
st2000 = pd.read_csv('../data/OEWS/2000state_dl.csv')
st2000.head(3)

area  st    state occ_code                         occ_titl  group  \
0     1  AL  Alabama  11-0000           Management Occupations  major   
1     1  AL  Alabama  11-1011                 Chief Executives    NaN   
2     1  AL  Alabama  11-1021  General and Operations Managers    NaN   

        tot_emp emp_prse   h_mean     a_mean  ... h_wpct75 h_wpct90 a_wpct10  \
0        120380      1.7    27.67      57560  ...    36.38    50.76    22530   
1          6090      3.7    48.26     100380  ...        #        #    43650   
2         38020      2.6    28.29      58840  ...    37.94    55.36    22230   

  a_wpct25   a_median a_wpct75 a_wpct90 annual release  year  
0    33310      51620    75680   105590    NaN     NaN  2000  
1    67190     107050        #        #    NaN     NaN  2000  
2    32540      50400    78920   115160    NaN     NaN  2000  

[3 rows x 24 columns]

In [14]:
st2000 = st2000[['area', 'st', 'state','occ_code', 'occ_titl', 'tot_emp', 'a_mean', 
                         'a_wpct10', 'a_wpct25', 'a_median', 'a_wpct75', 'a_wpct90']]
bad1 = ['*', '**', '#']
good1 = [0, 0, 145601]
st2000 = st2000.replace(to_replace=bad1, value=good1)
occ_codes_2000 = ['11-3021', '13-1081', '13-1111', '15-1011', '15-1021', '15-1031',                                                      
                                                    '15-1032', '15-1041', '15-1051', '15-1061', '15-1071', '15-1081',
                                                    '15-1099', '15-2031', '15-2041', '17-2061', '19-3021',
                                                    '43-9011', '43-9021', '43-9111', '49-2011']
st2000 = st2000.loc[st2000.occ_code.isin(occ_codes_2000)].reset_index(drop=True)
st2000 = st2000.loc[~st2000.state.isin(['Guam', 'Virgin Islands'])].reset_index(drop=True)
st2000['year'] = 2000
print(st2000.shape)
st2000.tail()


(950, 13)


area  st        state occ_code  \
945    72  PR  Puerto Rico  19-3021   
946    72  PR  Puerto Rico  43-9011   
947    72  PR  Puerto Rico  43-9021   
948    72  PR  Puerto Rico  43-9111   
949    72  PR  Puerto Rico  49-2011   

                                              occ_titl       tot_emp  \
945                           Market Research Analysts            90   
946                                 Computer Operators          1420   
947                                  Data Entry Keyers          3560   
948                             Statistical Assistants           170   
949  Computer, Automated Teller, and Office Machine...           790   

        a_mean a_wpct10 a_wpct25   a_median a_wpct75 a_wpct90  year  
945      40730    23350    28270      37870    49570    65810  2000  
946      19310    12290    14340      17220    22590    30300  2000  
947      15470    11570    12560      14260    17290    21490  2000  
948      16260    11340    12270      13820    18380    25770  2000  
949      22420    12330    14380      19680    27630    39370  2000

In [15]:
st2000 = st2000.rename(columns= {'occ_titl': 'occ_title','a_wpct10':'a_pct10', 'a_wpct25' : 'a_pct25', 
                                         'a_wpct75': 'a_pct75', 'a_wpct90': 'a_pct90'})
st2000.head(2)

area  st    state occ_code                                  occ_title  \
0     1  AL  Alabama  11-3021  Computer and Information Systems Managers   
1     1  AL  Alabama  13-1111                        Management Analysts   

        tot_emp     a_mean a_pct10 a_pct25   a_median a_pct75 a_pct90  year  
0          3090      63840   34110   44450      63920   80860   94230  2000  
1          2630      56600   23440   40120      53780   67840   90110  2000

In [16]:
st00_10_17 = st2017.append([st2010, st2000], ignore_index=True, verify_integrity=True, sort=False)
st00_10_17.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3057 entries, 0 to 3056
Data columns (total 15 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   area       3057 non-null   int64 
 1   st         3057 non-null   object
 2   state      3057 non-null   object
 3   occ_code   3057 non-null   object
 4   occ_title  3057 non-null   object
 5   tot_emp    3057 non-null   object
 6   jobs_1000  2107 non-null   object
 7   loc_q      2107 non-null   object
 8   a_mean     3057 non-null   object
 9   a_pct10    3057 non-null   object
 10  a_pct25    3057 non-null   object
 11  a_median   3057 non-null   object
 12  a_pct75    3057 non-null   object
 13  a_pct90    3057 non-null   object
 14  year       3057 non-null   int64 
dtypes: int64(2), object(13)
memory usage: 358.4+ KB


In [17]:
st00_10_17.to_csv('../data/OEWS/2000_10_17_stcombined.csv')